src/jit/flowgraph.cpp

```diff
@@ -4883,6 +4887,26 @@ void Compiler::fgFindJumpTargets(const BYTE* codeAddr, IL_OFFSET codeSize, BYTE*
 
     if (makeInlineObservations)
     {
+        if (retBlocks == 0)
+        {
+            // If there are no return blocks we know it does not return, however if there
+            // return blocks we don't know it returns as it may be counting unreachable code.
+            // So only make the CALLEE_DOES_NOT_RETURN observations if no returns.
+
+            if (!compInlineResult->UsesLegacyPolicy())
+            {
+                // Mark the call node as "no return" as it can impact caller's code quality.
+                impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+
+                if (compInlineResult->IsFailure())
```

Comment:
I can add an assert in `EnhancedLegacyPolicy::NoteBool`, but ideally would check it immediately after `fgMakeBasicBlocks`, but not sure I can inspect the observation there?